### PR TITLE
add doc-only `substrate` entry point crate

### PR DESCRIPTION
**Pull request by kianenigma**
Part of a series of PRs that I am getting ready in place for PBA. 

the new `substrate` crate is only meant to be used for very very high level documentation. This should be the crate that is shown when you also navigate to https://paritytech.github.io/substrate/master/

I will try and get a minimal version of #14137 ready soon as well, and instead of `frame_support`, we will use `frame` in here. This again makes me wish for #14268 to be the counter part of that as well. 

A similar crate in cumulus would also be a nice next step. 

Screenshot 2023-07-14 at 21 54 21

Screenshot 2023-07-14 at 21 54 29

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3193499

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3194935

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
+//! # Substrate
+//!
+//! Substrate is a Rust framework for building blockchains in a modular and extensible way. While in
+//! itself un-opinionated, it is the main engine behind the Polkadot ecosystem.
+//!
+//! [![github]](https://github.com/paritytech/substrate/) - [![polkadot]](https://polkadot.network)
+//!
+//! This crate in itself does not contain any code and is just meant ot be a documentation hub for
```
Comment (rphmeier) [inline src/lib.rs:8]: 'just meant to be documentation"

But also, why not re-export the main types and crates used in Substrate through this one? Is that the eventual plan?

Comment (rphmeier) [inline src/lib.rs:8]: 'just meant to be documentation"

But also, why not re-export the main types and crates used in Substrate through this one? Is that the eventual plan?

Comment (kianenigma) [reply src/lib.rs:8]: For FRAME, it is being made here: #14137. I imagine a companion crate called `substrate-client-builder` that is everything else that you might need, see #14268.

Once these two crates are in place, perhaps we can re-export them from this crate. 

As it stands now though, I find it hard to even narrow down what to re-export from this crate.

Comment (sam0x17) [reply src/lib.rs:8]: Yes I agree it would be a good idea to re-export `frame` here once it is ready

**Review by sam0x17 — APPROVED** [review]

**Review by gupnik — APPROVED** [review]

**Comment by kianenigma** [comment]
bot rebase

File: primitives/io/src/lib.rs [hunk 2]
```diff
@@ -15,7 +15,63 @@
...
-//! I/O host interface for substrate runtime.
+//! # Substrate Primitives: IO
+//!
+//! This crate contains interfaces for the runtime to communicate with the outside world, ergo `io`.
+//! In other context, such interfaces are referred to as "**host functions**".
+//!
+//! Each set of host functions are defined with an instance of the
+//! [`sp_runtime_interface::runtime_interface`] macro.
+//!
+//! Most notable, this crate contains host functions for:
```
Comment (gpestana) [inline primitives/io/src/lib.rs:26]:
```suggestion
//! Most notably, this crate contains host functions for:
```

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+//! upgrades out of the box. Combined with governance abilities that are shipped with `FRAME`, this
+//! enables a chain to always evolve.
+//!
+//! ## How to Get Stared
+//!
+//! Most developers want to leave the client side code as-is, and focus on the runtime. To do so,
+//! look into the [`frame_support`] crate, which is the entry point crate into runtime development.
+//!
+//! > Side note, it is entirely possible to craft a substrate-based runtime without FRAME, an
+//! > example of which can be found here.
```
Comment (gpestana) [inline src/lib.rs:41]: > example of which can be found here.

"here" where?

**Review by gpestana — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3200535

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,198 @@
+//! # Substrate
```
Comment (ggwpez) [inline src/lib.rs:1]: Do we need a License here?

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+//! This crate in itself does not contain any code and is just meant ot be a documentation hub for
+//! substrate-based crates.
+//!
+//! ## Overview
+//!
+//! Substrate approaches blockchain development with a focus on **safety**, **correctness**, and
+//! **upgradeability**.
+//!
+//! **Safety** is acquired through a use of Rust, a modern language empowering everyone to build
+//! reliable and efficient software.
```
Comment (ggwpez) [inline src/lib.rs:17]: Can you maybe explain a bit closer what is meant with `Safety` in the first place?  
I assume it alludes to panic-free and deterministic code that is easily auditable.

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+//!
+//! [![github]](https://github.com/paritytech/substrate/) - [![polkadot]](https://polkadot.network)
+//!
+//! This crate in itself does not contain any code and is just meant ot be a documentation hub for
+//! substrate-based crates.
+//!
+//! ## Overview
+//!
+//! Substrate approaches blockchain development with a focus on **safety**, **correctness**, and
+//! **upgradeability**.
```
Comment (ggwpez) [inline src/lib.rs:14]: I think its generality is also highly interesting. Substrate is using generic types quite often to make it possible to adapt to any use case.

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+//!
+//! **Safety** is acquired through a use of Rust, a modern language empowering everyone to build
+//! reliable and efficient software.
+//!
+//! **Correctness** is ensured through a rich type system enforcing semantic guarantees. This is
```
Comment (ggwpez) [inline src/lib.rs:19]: Correctness in which way? That it is easy to correctly implement some specific required runtime logic when using Substrate?

File: src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+//! To learn more about the substrate architecture using some visuals, see [`substrate_diagram`].
+//!
+//! All in all, this design enables all substrate-based chains to achieve forkless, self-enacting
+//! upgrades out of the box. Combined with governance abilities that are shipped with `FRAME`, this
+//! enables a chain to always evolve.
+//!
+//! ## How to Get Stared
+//!
+//! Most developers want to leave the client side code as-is, and focus on the runtime. To do so,
+//! look into the [`frame_support`] crate, which is the entry point crate into runtime development.
```
Comment (ggwpez) [inline src/lib.rs:38]: It will be `frame` in the future, or?

File: src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+//!
+//! > Side note, it is entirely possible to craft a substrate-based runtime without FRAME, an
+//! > example of which can be found [here](https://github.com/JoshOrndorff/frameless-node-template).
```
Comment (ggwpez) [inline src/lib.rs:41]: >Although possible, FRAME provides most primitives that a runtime developer will have to re-build manually. This comes at the cost of development time, maintenance cost and, possibly, correctness.

?

File: src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+//!   the client crates. Notable examples are crates such as [`sc-network`], various consensus
+//!   crates, [`sc-rpc-api`] and [`sc-client-db`], all of which are expected to reside in the client
+//!   side.
+//! * `sp` (short for *substrate-primitives*) crates, located under `./primitives` folder. These are
+//!   the traits that glue the client and runtime together, but are not opinionated about what
+//!   framework is using for building the runtime. Notable examples are [`sp-api`] and [`sp-io`],
+//!   which form the communication bridge between the client and runtime, as explained in
+//!   [`substrate_diagram`].
+//! * `pallet-*` and `frame-*` crates, located under `./frame` folder. These are the crates related
+//!   to FRAME. See [`frame_support`] for more information.
```
Comment (ggwpez) [inline src/lib.rs:70]: Some idea:  

Maybe mention the `utils` folder containing all kinds of helper binaries.  
So Substrate is in this way self contained and has little additional dependencies, besides `scale`, `type-info` and `bounded-collections`.

File: src/lib.rs (hunk 4 above)
Comment (kianenigma) [reply src/lib.rs:1]: yeah, I am going to use the license type used in the `sc-*` crates for now with version 0.0.0.

File: src/lib.rs (hunk 8 above)
Comment (kianenigma) [reply src/lib.rs:38]: Yup.

File: src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+//!
+//! ## Parachain?
+//!
+//! As noted above, Substrate is the main engine behind the Polkadot ecosystem. One of the ways
+//! through which Polkadot can be utilized is by building "parachains", blockchains that are
+//! connected to Polkadot's shared security.
+//!
+//! To build a parachain, one needs to use [`Cumulus`](https://github.com/paritytech/cumulus/), the
```
Comment (ggwpez) [inline src/lib.rs:78]: >one needs to use [`Cumulus`]

Similar to FRAME, it is not a hard requirement, but very convenient 😄

File: src/lib.rs (hunk 5 above)
Comment (kianenigma) [reply src/lib.rs:17]: I mostly want to lean into the type of memory safety guarantees that Rust brings us here.

File: src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+//! 	- [`sc-consensus-manual-seal`]
+//! 	- [`sc-consensus-pow`]
+//!
+//! Additional noteworthy external resources:
+//!
+//! - [Substrate Developer Hub](https://substrate.dev)
+//! - [Parity Tech's Documentation Hub](https://paritytech.github.io/)
+//! - [Frontier: Substrate's Ethereum Compatibility Library](https://paritytech.github.io/frontier/)
+//! - [Polkadot Wiki](https://wiki.polkadot.network/en/)
+//!
```
Comment (ggwpez) [inline src/lib.rs:101]: Maybe a section for interesting upstream deps like https://github.com/paritytech/parity-common and https://github.com/paritytech/parity-db?

File: src/lib.rs (hunk 7 above)
Comment (kianenigma) [reply src/lib.rs:19]: I meant to lean into the heavy use of the type system, which is truthfully more of a FRAME feature than substrate. I will revise the messaging a bit.

File: src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+/// 	end
+/// end
+/// ```
+///
+/// The client and the runtime of course need to communicate. This is done through two concepts:
+///
+/// 1. Host functions: a way for the (Wasm) runtime to talk to the client. All host functions are
+///    defined in [`sp-io`]. For example, [`sp-io::storage`] are the set of host functions that
+///    allow the runtime to read and write data to the on-chain state.
+/// 2. Runtime APIs: a way for the client to talk to the Wasm runtime. Runtime APIs are defined
```
Comment (ggwpez) [inline src/lib.rs:150]: A distinct difference between Runtime API and Host Function is that the Runtime API always creates a new WASM instance and does **not communicate** with an existing one.

**Review by ggwpez — APPROVED** [review]

**Comment by kianenigma** [comment]
@ggwpez I appreciate your criticism and it provoked me to rethink the whole story. 

Given your comment, I want to revise the messaging as such: 

For Substrate as a whole, our design philosophy is around: 

We acknowledge that 

1. technology and society evolves
2. humans are fallible

and therefore no single solution can remain relevant/correct for long. We counter this by: 

0. Using Rust, which helps limit human errors, while maintaining speed.
1. Being generic, which makes helps make the code more adaptable in the future. 
2. Upgradability through a meta-protocol.

At the FRAME level, we strive further for these goals via: 

0. Extensive use of type system and declarative code, in order to ensure correctness. 
1. Further generic and configurable design, which makes tweaking the system through governance possible

(needless to say, FRAME provides governance modules to enable the above).

File: src/lib.rs (hunk 13 above)
Comment (kianenigma) [reply src/lib.rs:150]: Seems more of an implementation detail rather than something worthy of being mentioned at this high level doc?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3202215

File: src/lib.rs (hunk 9 above)
Comment (kianenigma) [reply src/lib.rs:52]: Where is this coming from?

**Review by rcny — APPROVED** [review]

**Comment by kianenigma** [comment]
bot merge

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3211453

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3211911

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3220299

**Comment by kianenigma** [comment]
> DQ but this default-crate wont work anymore in the monorepo, or? Since we only have one top-level `Cargo.toml`...

This will be the crate at the root of `./substrate`. I presume the mono repo will have its own default crate at `.` that will just guide you toward either `substrate` or `cumulus`.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3221104

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3221355

**Comment by kianenigma** [comment]
bot merfe

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**